src/core/path.c

```diff
@@ -35,8 +35,8 @@ static int path_dispatch_io(sd_event_source *source, int fd, uint32_t revents, v
 int path_spec_watch(PathSpec *s, sd_event_io_handler_t handler) {
 
         static const int flags_table[_PATH_TYPE_MAX] = {
-                [PATH_EXISTS] = IN_DELETE_SELF|IN_MOVE_SELF|IN_ATTRIB,
-                [PATH_EXISTS_GLOB] = IN_DELETE_SELF|IN_MOVE_SELF|IN_ATTRIB,
+                [PATH_EXISTS] = IN_DELETE_SELF|IN_MOVE_SELF,
+                [PATH_EXISTS_GLOB] = IN_DELETE_SELF|IN_MOVE_SELF,
```

Comment:
this doesn't look right. we care about path deletion, i.e. the removal of a filename not the remove of the inode. and for that we need to watch the hardlink counter, which means IN_ATTRIB and not IN_DELETE_SELF. (THis is very confusing in inotify, and it took my a while to grok that...) Otherwise we are confused by multiple hardlinks to the same file and more importantly by apps keeping files open...

the logic must be: watch the the hardlink count, and when it changes test if the file is still in the path, and only act if it isn't anymore. We really don't care if the inode is actually removed or not, it's really just about whether it is linked into anything.

Reply:
Ack. I tested this now, and indeed my patch seems to break this. Ooooh, linux!